File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int target, int[] arr) {
        // pattern variable size sliding window 
        int low = 0;
        int high = 0;
        int res = Integer.MAX_VALUE;
        int sum = 0;
        int n = arr.length;

        while(high < n){
            //hire
            sum = sum + arr[high];
            while(sum>=target){
                int len = high - low + 1;
                res = Math.min(res,len);
                // fire 
                sum =  sum - arr[low];
                low++;
            }
            high++;
        }
        return res == Integer.MAX_VALUE ? 0 : res;
        
    }
}

File: 0209-minimum-size-subarray-sum/0209-minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int minLen = INT_MAX;
        int left = 0;
        int currSum = 0;

        for(int right = 0 ; right < nums.size() ; right++){
            currSum += nums[right];
            while(currSum >= target){
                if(right-left+1 < minLen){
                    minLen = right-left+1;
                }
                currSum -=nums[left];
                left++;
            }
        }
        return minLen != INT_MAX ? minLen : 0;
        
    }
};